New Nginx Tracing Module (#15019)
DataDog/documentation: content/en/tracing/trace_collection/proxy_setup/_index.md

- Nginx as an Ingress Controller for Kubernetes.

## Nginx with Datadog module
Datadog provides an nginx module for distributed tracing. This module is the
preferred way to add Datadog distributed tracing to nginx.

### Module installation
There is one version of the Datadog nginx module for each [nginx Docker image
tag][13]. Install the module by downloading the appropriate file from the
[latest nginx-datadog GitHub release][14] and extracting it into Nginx's
modules directory.

For example, if nginx version 1.23.1 is running on a Debian-based system, then
the appropriate nginx image tag is [1.23.1][15]. The corresponding Alpine-based
image is tagged [1.23.1-alpine][16].

```bash
get_latest_release() {
  curl --silent "https://api.github.com/repos/$1/releases/latest" | jq --raw-output .tag_name
}
NGINX_IMAGE_TAG=1.23.1
RELEASE_TAG=$(get_latest_release DataDog/nginx-datadog)
tarball="$NGINX_IMAGE_TAG-ngx_http_datadog_module.so.tgz"
wget "https://github.com/DataDog/nginx-datadog/releases/download/$RELEASE_TAG/$tarball"
tar -xzf "$tarball" -C /usr/lib/nginx/modules
rm "$tarball"
```

### Nginx configuration with Datadog module

> In the topmost section of the nginx configuration, load the Datadog module.

Reviewer: ```suggestion
In the topmost section of the Nginx configuration, load the Datadog module.
```

```nginx
load_module modules/ngx_http_datadog_module.so;
```

The default configuration connects to a local Datadog Agent and produces traces
for all Nginx locations. Specify custom configuration in a `datadog` JSON block
within the `http` section of the nginx configuration.

> 

Reviewer 2: ```suggestion
As an example, to set a service name to `usage-internal-nginx` and sample 10% of your traces:
```

```nginx
load_module modules/ngx_http_datadog_module.so;

http {
  datadog {

>     "service": "usage-internal-nginx",

Reviewer 2: Is service part of the documentation? (https://github.com/DataDog/nginx-datadog/blob/master/doc/API.md#datadog)

Author: yes

    "sample_rate": 0.1
  }
}
```

For information about fields supported by the `datadog` directive and about
other configuration directives supported by the module, see the [API
documentation][17].

> 

Reviewer 2: Should we add the evaluation order between directive and environment variables?

Author: That's covered in the documentation of the environment variables, which is linked in the API documentation.

{{% /tab %}}
{{% tab "Nginx" %}}

Datadog APM supports Nginx in multiple configurations:
- Nginx operated as a proxy with tracing provided by the new Datadog module.
- Nginx operated as a proxy with tracing provided by the OpenTracing module.
- Nginx as an Ingress Controller for Kubernetes.

## Nginx with Datadog module

> Datadog provides an Nginx module for distributed tracing. This module is 
> currently in beta.

Reviewer 2: ```suggestion
<div class="alert alert-warning">
Datadog Nginx module is currently in beta. <a href="https://docs.datadoghq.com/help/">Reach out to support</a> if you have any feedback.
</div>
```

}
```

> If no sample rate is specified, the [Datadog Agent calculated sampling rates][7] (10 traces per second per Agent) are applied.

Reviewer 2: Does it also apply to Nginx module? As this section falls under the Nginx with opentracing module I would assume it is only relevant for this case

Author: `### Nginx Sampling` is an `h3` section, the same as the OpenTracing and Datadog module sections.  So, there is one section about getting the Datadog module, another about getting the OpenTracing module, and then a third about configuring sampling for them.

This is possibly confusing.  What might be better?

Author: I am wrong.  Will fix.

[10]: /getting_started/tagging/unified_service_tagging/?tab=kubernetes#configuration-1
[11]: /tracing/setup/cpp/#environment-variables
[12]: https://istio.io/docs/ops/configuration/traffic-management/protocol-selection/#manual-protocol-selection
[13]: https://hub.docker.com/_/nginx/tags
[14]: https://github.com/DataDog/nginx-datadog/releases/latest
[15]: https://hub.docker.com/layers/nginx/library/nginx/1.23.1/images/sha256-f26fbadb0acab4a21ecb4e337a326907e61fbec36c9a9b52e725669d99ed1261?context=explore
[16]: https://hub.docker.com/layers/nginx/library/nginx/1.23.1-alpine/images/sha256-2959a35e1b1e61e2419c01e0e457f75497e02d039360a658b66ff2d4caab19c4?context=explore
[17]: https://github.com/DataDog/nginx-datadog/blob/master/doc/API.md

> [18]: https://github.com/DataDog/nginx-datadog/blob/master/doc/API.md#datadog

Author: Ah, that explains it.  Thanks.
